### PR TITLE
update rust to 1.60, solana to 1.11.6

### PR DESCRIPTION
**Pull request by 2501babe**
i marked two things in the stake client as `#[allow(deprecated)]` because after looking at the relevant types in `solana-client` i dont understand what the correct change is

```
warning: use of deprecated field `solana_client::rpc_filter::Memcmp::encoding`: Field has no server-side effect. Specify encoding with `MemcmpEncodedBytes` variant instead.
  --> stake-pool/cli/src/client.rs:91:21
   |
91 |                     encoding: None,
   |                     ^^^^^^^^^^^^^^
   |
   = note: `#[warn(deprecated)]` on by default

warning: use of deprecated field `solana_client::rpc_filter::Memcmp::encoding`: Field has no server-side effect. Specify encoding with `MemcmpEncodedBytes` variant instead.
   --> stake-pool/cli/src/client.rs:136:21
    |
136 |                     encoding: Some(MemcmpEncoding::Binary),
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

if possible i wanna merge this as is and worry about that later tho

### CODE REVIEW

File: stake-pool/cli/src/client.rs [hunk 1]
```diff
@@ -84,6 +84,7 @@ pub(crate) fn get_stake_pools(
     rpc_client
         .get_program_accounts_with_config(
             &spl_stake_pool::id(),
+            #[allow(deprecated)] // TODO figure out what encoding to use
```
Comment (2501babe) [inline stake-pool/cli/src/client.rs:87]: this

File: stake-pool/cli/src/client.rs [hunk 2]
```diff
@@ -127,6 +128,7 @@ pub(crate) fn get_all_stake(
 ) -> Result<HashSet<Pubkey>, ClientError> {
     let all_stake_accounts = rpc_client.get_program_accounts_with_config(
         &stake::program::id(),
+        #[allow(deprecated)] // TODO figure out what encoding to use
```
Comment (2501babe) [inline stake-pool/cli/src/client.rs:131]: and this

**Review by joncinque — DISMISSED** [review]
Looks good! Could you just add an issue to investigate and fix those `#[allow(deprecated)]`s? Might even be a possible "good-first-issue"

**Comment by CriesofCarrots** [comment]
> i dont understand what the correct change is

(Sorry for the flyby) The correct change should be to use a constructor fn for Memcmp: https://github.com/solana-labs/solana/blob/225cddcffbf453cc05f5c6ca2da74509a3fcde07/client/src/rpc_filter.rs#L147-L161
At a glance, both should probably be `Memcmp::new_raw_bytes()`, although I can take closer look if helpful.

**Comment by 2501babe** [comment]
re: the failures, it looks like the flags for one of the upstream tools (either `cargo-test-bpf` or `cargo-test-sbf`) have changed... or, rather, the flags that one of them generates to pass to something else are being rejected by the recipient because perhaps the call chain has changed?

```
    Finished test [unoptimized + debuginfo] target(s) in 5m 29s
     Running unittests (/home/runner/work/solana-program-library/solana-program-library/target/debug/deps/spl_feature_proposal-2570cef13791f5fb)
error: Unrecognized option: 'arch'
error: test failed, to rerun pass '--lib'
cargo-test-sbf child: /home/runner/.local/share/solana/install/active_release/bin/cargo-build-sbf --manifest-path /home/runner/work/solana-program-library/solana-program-library/feature-proposal/program/Cargo.toml --sbf-out-dir /home/runner/work/solana-program-library/solana-program-library/target/deploy --arch sbf
cargo-test-sbf child: /home/runner/.rustup/toolchains/1.60.0-x86_64-unknown-linux-gnu/bin/cargo test --manifest-path /home/runner/work/solana-program-library/solana-program-library/feature-proposal/program/Cargo.toml --features test-bpf -- --nocapture --arch bpf

Error: Process completed with exit code 1.
```

@dmakarov do you have any insight into this?

**Comment by dmakarov** [comment]
> re: the failures, it looks like the flags for one of the upstream tools (either `cargo-test-bpf` or `cargo-test-sbf`) have changed... or, rather, the flags that one of them generates to pass to something else are being rejected by the recipient because perhaps the call chain has changed?
> 
> ```
>     Finished test [unoptimized + debuginfo] target(s) in 5m 29s
>      Running unittests (/home/runner/work/solana-program-library/solana-program-library/target/debug/deps/spl_feature_proposal-2570cef13791f5fb)
> error: Unrecognized option: 'arch'
> error: test failed, to rerun pass '--lib'
> cargo-test-sbf child: /home/runner/.local/share/solana/install/active_release/bin/cargo-build-sbf --manifest-path /home/runner/work/solana-program-library/solana-program-library/feature-proposal/program/Cargo.toml --sbf-out-dir /home/runner/work/solana-program-library/solana-program-library/target/deploy --arch sbf
> cargo-test-sbf child: /home/runner/.rustup/toolchains/1.60.0-x86_64-unknown-linux-gnu/bin/cargo test --manifest-path /home/runner/work/solana-program-library/solana-program-library/feature-proposal/program/Cargo.toml --features test-bpf -- --nocapture --arch bpf
> 
> Error: Process completed with exit code 1.
> ```
> 
> @dmakarov do you have any insight into this?

I don't know how cargo-test-sbf was started. Any options that appear on `cargo test` command line after `--features test-bpf` were added to cargo-test-sbf command line after `--`. cargo-test-sbf simply forwards these options to `cargo test` command line. Naturally test harness rejects `--arch` command line option.

File: stake-pool/cli/src/client.rs [hunk 3]
```diff
@@ -128,15 +127,13 @@ pub(crate) fn get_all_stake(
...
             filters: Some(vec![
                 // Filter by `Meta::authorized::staker`, which begins at byte offset 12
-                RpcFilterType::Memcmp(Memcmp {
-                    offset: 12,
-                    bytes: MemcmpEncodedBytes::Base58(authorized_staker.to_string()),
-                    encoding: Some(MemcmpEncoding::Binary),
-                }),
+                RpcFilterType::Memcmp(Memcmp::new_base58_encoded(
+                    12,
+                    &authorized_staker.to_bytes(),
```
Comment (CriesofCarrots) [inline stake-pool/cli/src/client.rs:135]:
```suggestion
                    authorized_staker.as_ref(),
```

File: stake-pool/cli/src/client.rs [hunk 4]
```diff
@@ -84,13 +84,12 @@ pub(crate) fn get_stake_pools(
...
             RpcProgramAccountsConfig {
-                filters: Some(vec![RpcFilterType::Memcmp(Memcmp {
-                    offset: 0, // 0 is the account type
-                    bytes: MemcmpEncodedBytes::Base58("2".to_string()),
-                    encoding: None,
-                })]),
+                // 0 is the account type
+                filters: Some(vec![RpcFilterType::Memcmp(Memcmp::new_raw_bytes(
+                    0,
+                    vec![1],
```
Comment (2501babe) [inline stake-pool/cli/src/client.rs:91]: i couldnt find tests for this program but i confirmed externally this should be right

```
use solana_program::pubkey::Pubkey;
use solana_client::rpc_filter::{Memcmp, MemcmpEncoding, MemcmpEncodedBytes};

fn main() {
    let key = Pubkey::new_rand();

    let first_t = Memcmp {
        offset: 0,
        bytes: MemcmpEncodedBytes::Base58("2".to_string()),
        encoding: None,
    };

    let first_o = Memcmp::new_raw_bytes(0, vec![1]);

    let second_t = Memcmp {
        offset: 12,
        bytes: MemcmpEncodedBytes::Base58(key.to_string()),
        encoding: Some(MemcmpEncoding::Binary),
    };

    let second_o = Memcmp::new_base58_encoded(12, &key.to_bytes());

    println!("first (theirs):  {:?}", first_t.bytes());
    println!("first (ours):    {:?}", first_o.bytes());
    println!("\nsecond (theirs): {:?}", second_t.bytes());
    println!("second (ours):   {:?}", second_o.bytes());
}
```

```
    Finished dev [unoptimized + debuginfo] target(s) in 13.13s
     Running `target/debug/memcmp-test`
first (theirs):  Some([1])
first (ours):    Some([1])

second (theirs): Some([172, 202, 17, 192, 239, 120, 241, 109, 190, 67, 90, 45, 39, 45, 102, 1, 126, 172, 5, 167, 168, 222, 203, 72, 209, 138, 25, 102, 201, 25, 217, 23])
second (ours):   Some([172, 202, 17, 192, 239, 120, 241, 109, 190, 67, 90, 45, 39, 45, 102, 1, 126, 172, 5, 167, 168, 222, 203, 72, 209, 138, 25, 102, 201, 25, 217, 23])
```

**Comment by 2501babe** [comment]
the issue with `--arch bpf` is something in the sdk utils is tacking it on at the end in 1.16 where this didnt happen in 1.15. per jon i tried switching from test-bpf calls to test-sbf calls, which doesnt have this wrapper layer

i think we will still have a few test failures now (im not sure how to run these job on my system and the tests pass when invoked normally so i have to rely on ci) which could be 1.16 changes still to triage

**Comment by dmakarov** [comment]
> the issue with `--arch bpf` is something in the sdk utils is tacking it on at the end in 1.16 where this didnt happen in 1.15. per jon i tried switching from test-bpf calls to test-sbf calls, which doesnt have this wrapper layer

@joncinque just fixed this in #27221

> (im not sure how to run these job on my system and the tests pass when invoked normally so i have to rely on ci)

Could you run the CI script locally?

```
./ci/cargo-test-bpf.sh token
```

File: ci/cargo-test-bpf.sh [hunk 5]
```diff
@@ -31,7 +31,7 @@ if [[ -r $run_dir/Cargo.toml ]]; then
     # Build/test just one BPF program
     set -x
     cd $run_dir
-    cargo +"$rust_stable" test-sbf -- --nocapture
+    cargo +"$rust_stable" test-sbf --arch bpf -- --nocapture
```
Comment (dmakarov) [inline ci/cargo-test-bpf.sh:34]: You don't need to specify bpf architecture. It's the default. The option was added to test sbfv2 architecture as means to test the changes in the new architecture.

**Comment by 2501babe** [comment]
after fighting this awhile the two outstanding issues are:
* token tests fail when `--arch bpf` is removed from the ci script. after going through logs, it looks like the reason this seemed to fix the failures is because it causes the tests that happen to fail to not run at all. i dont know why
* stake pool python client tests fail. i thought this was due to my change to the stake pool cli, but it still fails after reverting it. possibly other things changed with the filters, possibly unrelated. ive been unable to run the python stuff locally

ill try again tomorrow

**Comment by dmakarov** [comment]
> after fighting this awhile the two outstanding issues are:
> 
>     * token tests fail when `--arch bpf` is removed from the ci script. after going through logs, it looks like the reason this seemed to fix the failures is because it causes the tests that happen to fail to not run at all. i dont know why
> 
>     * stake pool python client tests fail. i thought this was due to my change to the stake pool cli, but it still fails after reverting it. possibly other things changed with the filters, possibly unrelated. ive been unable to run the python stuff locally
> 
> 
> ill try again tomorrow

Adding `--arch bpf` enables these functions to be compiled https://github.com/solana-labs/solana-program-library/blob/8d88fb6c88a1bd5d0b55a47690943a173d7dd86e/token/program/src/processor.rs#L1023-L1027 and https://github.com/solana-labs/solana-program-library/blob/8d88fb6c88a1bd5d0b55a47690943a173d7dd86e/token/program-2022/src/processor.rs#L1403-L1407

These `#cfg[]` annotations should be changed to `#[cfg(target_os = "solana")]`, similarly `#[cfg(not(target_arch = "bpf"))]` should be changed to `#[cfg(not(target_os = "solana"))]`. Then cargo-build-sbf will do the right thing without the option `--arch bpf`.

**Comment by joncinque** [comment]
> These #cfg[] annotations should be changed to #[cfg(target_os = "solana")], similarly #[cfg(not(target_arch = "bpf"))] should be changed to #[cfg(not(target_os = "solana"))]. Then cargo-build-sbf will do the right thing without the option --arch bpf.

That's my fault, I made those changes in a security advisory before we had `target_os` and didn't think to update them when I finally merged it.  I put in #3493 to address that bit for you

**Comment by joncinque** [comment]
@2501babe let me look into the stake pool failures, it might be due to the minimum stake delegation

**Comment by joncinque** [comment]
I fixed up the tests, rebased, and reverted your revert of the change, since you did it correctly. I ran a little test against testnet and it was perfect

**Comment by joncinque** [comment]
I decided to just go the whole way and remove "bpf" everywhere in the code, so our features and everything are aligned now

File: stake-pool/py/tests/test_bot_rebalance.py [hunk 6]
```diff
@@ -19,7 +19,10 @@ async def test_rebalance_this_is_very_slow(async_client, validators, payer, stak
     (stake_pool_address, validator_list_address) = stake_pool_addresses
     resp = await async_client.get_minimum_balance_for_rent_exemption(STAKE_LEN)
     stake_rent_exemption = resp['result']
-    increase_amount = MINIMUM_ACTIVE_STAKE
+    # With minimum delegation at MINIMUM_DELEGATION + rent-exemption, when
+    # decreasing, we'll need rent exemption + minimum delegation delegated to
+    # cover all movements
+    increase_amount = MINIMUM_ACTIVE_STAKE + stake_rent_exemption
```
Comment (2501babe) [inline stake-pool/py/tests/test_bot_rebalance.py:25]: ohhh it was the 1.11 change that everything needs to be rent exempt, i understand now

**Comment by 2501babe** [comment]
thank you everyone for your help!!

**Pull request merged**